### ISSUE TITLE
Replace domains -> types, .inputs+.output -> type hints

### ISSUE THREAD

**Issue by fritzo** [pyro-ppl/funsor#351]
This issue proposes to replace funsor `Domain` objects with new subscripted types a la Python 3's `typing` module.

The goal is to make Funsor more Pythonic and to make funsors act more like Python functions with support pointwise operations. In particular many of our design questions about multiple inputs or outputs could be resolved by "following Python". A simple example of more-Pythonic syntax is:
```py
# decorator syntax to read .inputs and .outputs from a type annotation
@to_funsor
def f(x: Reals[3], y: Reals[3,3], t: Reals[()]) -> Reals[3]:
    ...

# explicitly specify type as Callable[..., ...]
g = to_funsor(lambda x: x.sum(), Callable[[Reals[3]], Reals[()])
    ...

h = f + g  # materialized view of lambda x, y, t: f(x, y, t) + g(x)
```

## Tasks
- [ ] Refactor domains to be typing-compatible type objects: `Bint[n]`, `Reals[()]`, `Reals[m,n]`, ...
- [ ] Update interfaces to make use of typing-compatible types (e.g. replace `function` with `to_funsor` of a callable)
- [ ] Possibly promoting `Funsor` subytype themselves to subtypes of `Callable`, as suggested by @eb8680 (though this may warrant a separate issue of larger scope)

**Comment by eb8680** [comment]
This definitely seems like the right way to go.  Some design questions:
- Could we replace `find_domain` with some standard Python type inference library?  Are there existing special array types in NumPy or elsewhere we should use rather than making a custom `Domain`?
- Would some version of this also get us shape polymorphism (as in #214) for free?  If not, does it at least not put up an intrinsic barrier to shape variables ([`typing.Literal` is currently very limited](https://www.python.org/dev/peps/pep-0586/#interactions-with-generics))? 
- How would this interact with surrounding typed Python code?  That is, what is the type signature of `to_funsor`?
- Should we distinguish between a black-box `funsor.function` that creates a custom new `funsor.Function` term and a transparent `funsor.to_funsor` that attempts to extract a funsor expression by evaluating the function on appropriately typed `funsor.Variable`s?

>  Possibly promote Funsor types themselves to subtypes of Callable

I was more thinking about making all `funsor.Funsor`s subtypes of `typing.Generic`, so that the parametrized types generated by `FunsorMeta` and used in our interpreters would be proper Python types and we could typecheck our interpreters with standard Python typecheckers.  That seems orthogonal to this issue - it is about the types of Funsor *interpreters*, rather than the types of Funsors themselves as in this issue.

**Comment by fritzo** [comment]
@eb8680 great questions!
1. I like the idea of being not only more Pythonic but closer to the NumPy ecosystem. Let's see what we can find.
2. Yes I believe we can get polymorphism using type variables 🎉 
3. Maybe we can make `Funsor` a subclass of `Callable`, i.e. a callable that supports pointwise evaluation.
4. I suppose `to_funsor()` behavior could depend on interpretation...

**Comment by fritzo** [comment]
@eb8680 suggested this will be easier after #491 , e.g. we could add an `@ops.make` decorator and use it in the VAE example e.g.

<details>

```py
...
encoder = Encoder()
decoder = Decoder()

# Version 0. current state
encode = funsor.function(Reals[28, 28], (Reals[20], Reals[20]))(encoder)
decode = funsor.function(Reals[20], Reals[28, 28])(decoder)

# Version 1. replace Function with and Op
@UnaryOp.make
def encode(image):
    return encoder(image)

@find_domain.register(encode)
def _(op, image):
    return typing.Tuple[Reals[20], Reals[20]]

# Version 2. read signature from the nn.Module
encode = UnaryOp.make(encoder)
# ...find_domain.register

# Version 3. registers find_domain based on type annotations
@UnaryOp.make
def encode(image: Reals[28, 28]) -> typing.Tuple[Reals[20], Reals[20]]:
    return encoder(image)

# Version 4. automatically determines arity
@ops.make(arity=1)  # manually specify
@ops.make  # assume arity = nargs
def encode(image: Reals[28, 28]) -> typing.Tuple[Reals[20], Reals[20]]:
    return encoder(image)
```

</details>
